File: java/src/main/java/br/com/emendes/problems/TextJustification.java
package br.com.emendes.problems;

import java.util.ArrayList;
import java.util.List;

/**
 * Da uma array de Strings e uma largura maxWidth, formate o texto de maneira que cada linha tenha
 * exatamente maxWidth caracteres e totalmente justificado (direita e esquerda).<br><br>
 * <p>
 * Você deve embalar suas palavras de uma forma gananciosa, isso é, coloque o máximo de palavras que puder em cada linha.
 * Preencha espaços extras ' ' quando necessário para que cada linha tenha exatamente caracteres maxWidth.<br><br>
 * <p>
 * Os espaços extras entre as palavras devem ser distribuídos da forma mais uniforme possível.
 * Se o número de espaços em uma linha não for dividido igualmente entre as palavras,
 * os slots vazios à esquerda receberão mais espaços do que os slots à direita.<br><br>
 * <p>
 * Para a última linha do texto, deve ser justificado à esquerda e nenhum espaço extra deve ser
 * inserido entre as palavras.<br><br>
 * <p>
 * Restrições:<br>
 * 1 <= words.length <= 300<br>
 * 1 <= words[i].length <= 20<br>
 * words[i] consiste apenas de letras inglesas e símbolos.<br>
 * 1 <= maxWidth <= 100<br>
 * words[i].length <= maxWidth
 */
public class TextJustification {

  private int currIndex = 0;

  public List<String> fullJustify(String[] words, int maxWidth) {
    currIndex = 0;
    List<String> answer = new ArrayList<>();

    while (currIndex < words.length) {
      answer.add(justifyLine(words, maxWidth));
    }

    return answer;
  }

  private String justifyLine(String[] words, int maxWidth) {
    int[] spaces = new int[60];

    int currWidth = words[currIndex].length();
    int start = currIndex;
    int end = currIndex;
    int spaceIndex = 1;

    currIndex++;
    while (currIndex < words.length && words[currIndex].length() + currWidth < maxWidth) {
      spaces[spaceIndex - 1]++;
      currWidth += 1 + words[currIndex].length();

      end++;
      spaceIndex++;
      currIndex++;
    }

    // caso seja a última linha.
    if (currIndex == words.length) {
      spaces[spaceIndex - 1] = maxWidth - currWidth;
    } else if (words[currIndex].length() + currWidth >= maxWidth) {
      int totalAdditionalSpace = maxWidth - currWidth;

      if (start == end) spaces[0] = totalAdditionalSpace;
      else {
        while (totalAdditionalSpace > 0) {
          for (int i = 0; i < end - start && totalAdditionalSpace > 0; i++) {
            spaces[i]++;
            totalAdditionalSpace--;
          }
        }
      }
    }

    return append(words, start, end, spaces);
  }

  private String append(String[] words, int start, int end, int[] spaces) {
    StringBuilder line = new StringBuilder();

    for (int i = 0; i <= end - start; i++) {
      line.append(words[start + i]);
      line.append(" ".repeat(spaces[i]));
    }

    return line.toString();
  }

}

File: javascript/problems/TextJustification.js
/**
 * Dado um array de strings words e uma largura maxWidth, formate o texto de modo que cada linha
 * tenha exatamente maxWidth caracteres e seja totalmente justificada (esquerda e direita).
 *
 * Você deve colocar suas palavras de forma gananciosa; ou seja,
 * colocar o máximo de palavras possível em cada linha. Insira espaços extras ' '
 * quando necessário para que cada linha tenha exatamente maxWidth caracteres.
 *
 * Espaços extras entre palavras devem ser distribuídos o mais uniformemente possível.
 * Se o número de espaços em uma linha não for dividido uniformemente entre as palavras,
 * os slots vazios à esquerda receberão mais espaços do que os slots à direita.
 *
 * A última linha do texto deve ser justificada à esquerda e
 * nenhum espaço extra deve ser inserido entre as palavras.
 *
 * Nota:
 *
 * Uma palavra é definida como uma sequência de caracteres que consiste apenas de caracteres sem espaços.
 *
 * O comprimento de cada palavra é garantido como maior que 0 e não excede maxWidth.
 *
 * A matriz de entrada words contém pelo menos uma palavra.
 *
 * Restrições:
 *
 * 1 <= words.length <= 300
 *
 * 1 <= words[i].length <= 20
 *
 * words[i] consiste de letras e símbolos.
 *
 * 1 <= maxWidth <= 100
 *
 * words[i].length <= maxWidth
 *
 * @param {string[]} words
 * @param {number} maxWidth
 * @return {string[]}
 */
const fullJustify = function (words, maxWidth) {
  let currWidth = 0;
  let nextLine = [];
  const justifiedText = [];
  for (const word of words) {
    if (currWidth + word.length < maxWidth) {
      nextLine.push(word);
      currWidth += word.length + 1;
    } else if (currWidth + word.length === maxWidth) {
      nextLine.push(word);
      justifiedText.push(nextLine.join(" "));
      currWidth = 0;
      nextLine = [];
    } else {
      justifiedText.push(createLine(nextLine, currWidth - 1, maxWidth));
      nextLine = [word];
      currWidth = word.length + 1;
    }
  }
  if (nextLine.length > 0) {
    let lastLine = nextLine.join(" ") + " ".repeat(maxWidth - currWidth + 1);
    justifiedText.push(lastLine);
  }
  return justifiedText;
};

const createLine = (words, currWidth, maxWidth) => {
  if (words.length === 1) return words[0] + " ".repeat(maxWidth - currWidth);
  const length = words.length - 1;
  const minSpace = 1 + Math.floor((maxWidth - currWidth) / length);
  const spaces = new Array(length).fill(minSpace);
  if (minSpace > 1) currWidth += length * (minSpace - 1);
  let i = 0;
  while (currWidth < maxWidth) {
    spaces[i++]++;
    if (i === spaces.length) i = 0;
    currWidth++;
  }
  let line = words[0];
  i = 1;
  let j = 0;
  while (i < words.length) {
    line += " ".repeat(spaces[j++]) + words[i++];
  }
  return line;
};
